packages/foundry/contracts/test/other/tether/TetherJoin.t.sol

```diff
@@ -0,0 +1,285 @@
+        join.join(address(join), unclaimedTokens);
+    }  
+}
+
+contract DeployedTest is Deployed {
+
+    function testJoinAuth() public {
+        vm.expectRevert("Access denied");
+        vm.prank(user);
+        join.join(user, unit);
+    }
+
+    function testExitAuth() public {
+        vm.expectRevert("Access denied");
+        vm.prank(user);
+        join.exit(user, unit);
+    }
+
+    function testRetrieveAuth() public {
+        vm.expectRevert("Access denied");
+        vm.prank(user);
+        join.retrieve(otherToken, other);
+    }
+    
+    function testJoinPull() public {
+        track("userBalance", tether.balanceOf(user));
+        track("storedBalance", join.storedBalance());
+        track("joinBalance", tether.balanceOf(address(join)));
+
+        vm.prank(user);
+        tether.approve(address(join), unit);
+        vm.prank(ladle);
+        join.join(user, unit);
+
+        assertTrackMinusEq("userBalance", unit, tether.balanceOf(user));
+        assertTrackPlusEq("storedBalance", unit, join.storedBalance());
+        assertTrackPlusEq("joinBalance", unit, tether.balanceOf(address(join)));
+    }
+
+    function testJoinPush() public {
+        track("userBalance", tether.balanceOf(user));
+        track("storedBalance", join.storedBalance());
+        track("joinBalance", tether.balanceOf(address(join)));
+
+        vm.prank(user);
+        tether.transfer(address(join), unit);
+        vm.prank(ladle);
+        join.join(user, unit);
+
+        assertTrackMinusEq("userBalance", unit, tether.balanceOf(user));
+        assertTrackPlusEq("storedBalance", unit, join.storedBalance());
+        assertTrackPlusEq("joinBalance", unit, tether.balanceOf(address(join)));
+    }
+
+    function testJoinCombine() public {
+        track("userBalance", tether.balanceOf(user));
+        track("storedBalance", join.storedBalance());
+        track("joinBalance", tether.balanceOf(address(join)));
+
+        vm.prank(user);
+        tether.approve(address(join), unit / 2);
+        vm.prank(user);
+        tether.transfer(address(join), unit / 2);
+        vm.prank(ladle);
+        join.join(user, unit);
+
+        assertTrackMinusEq("userBalance", unit, tether.balanceOf(user));
+        assertTrackPlusEq("storedBalance", unit, join.storedBalance());
+        assertTrackPlusEq("joinBalance", unit, tether.balanceOf(address(join)));
+    }
+}
+
+abstract contract WithFees is Deployed {
+    function setUp() public override virtual {
+        super.setUp();
+
+        // enable fees
+        vm.prank(tetherMultiSig);
+        tether.setParams(19, 49);    // maximum
+    }
+}
+
+contract WithFeesTest is WithFees {
+    function testJoinPullWithFees() public {
+        track("userBalance", tether.balanceOf(user));
+        track("storedBalance", join.storedBalance());
+        track("joinBalance", tether.balanceOf(address(join)));
+
+        uint256 units = unit * 5;
+        uint256 fee = tether.basisPointsRate() * 100;
```

Comment:
Is this right?

You are using `fee` as a decimal factor, to multiply amounts by. Below you have `units * unit / (unit - fee)`, so `fee` needs to have the same decimals as `unit`. That means that you need to convert from `basisPointsRate` which is an fp4 (fixed point with 4 decimals) to `unit` which is a fixed point with `tether.decimals()`.

Wouldn't it be:
`uint256 fee = tether.basisPointsRate() * 10000;`